### DISCUSSION TITLE
Control the variables available to the operator?

### DISCUSSION THREAD

**Discussion by Zhonglu** [MilesCranmer/PySR#258] [Q&A] [answered]
We can see the that now we can constrain the operators used by operators, which is great!

However, it seems we cannot control the variables used by operators. For example, we might know by priori that X_0 can only by used by multiply as  * X_0, and X_0 only appears in a certain term.

In other words, is it possible to predetermine a function, leaving a few parameters for pysr for fitting?

For example, y = a * x0^b * x1 + c * x2, and we may only need pysr to find out a, b and c.

Is there any template to try out? Should we try the penalty one?

Appreciate any comment,

Mike

**Comment by Zhonglu** [comment]
More fundamentally, what would be the best practice to incorporate known knowledge into the symbolic regression process?

**Comment by MilesCranmer** [comment, marked as answer]
For this, and more generally for other constraints which are very specific, it might be best to write a custom `eval_loss` function, as described here: https://astroautomata.com/PySR/backend/.

---

Update: you can now control this with `full_objective` (PySR) or `loss_function` (SymbolicRegression.jl).
